Cluster Certificate Management (#1308)
spectrocloud/librarium: content/docs/04-clusters/06-cluster-management/02.5-certificate-management.md

fullWidth: false
---

import Tabs from 'shared/components/ui/Tabs';
import WarningBox from 'shared/components/WarningBox';
import InfoBox from 'shared/components/InfoBox';

# Overview

Palette installs Kubernetes through the tool, [kubeadm](https://kubernetes.io/docs/reference/setup-tools/kubeadm). As a result, all deployed clusters include auto-generated Public Key Infrastructure (PKI) certificates created by kubeadm. We recommend you review the [PKI certificates and requirement](https://kubernetes.io/docs/setup/best-practices/certificates) documentation from the official Kubernetes documentation to learn more about the auto-generated certificates and to understand their purpose better.

This reference page focuses on how to renew the PKI certificates through Palette. You have two options for how you can renew the cluster PKI certificates:

  <br />

  * Automatic Certificate Renewal

  * Manual Certificate Renewal

<InfoBox>

Certificates created by kubeadm expire after 365 days. The Root Certificate Authority (CA) is valid for 3652 days or ten years.

</InfoBox>

You can learn more about each option in the following sections. 

# Automatic Certificate Renewal

> When you upgrade the control plane on a cluster, the PKI certificates are automatically updated. Upgrading a cluster's Kubernetes version, whether a minor patch or a major release, results in renewed PKI certificates. This is the method that requires the least user actions when it comes to renewing PKI certificates. We recommend regularly updating your clusters to stay up-to-date with security fixes and best practices. By keeping your host cluster up-to-date, you prevent the scenario of PKI certificates from reaching their one-year expiration date.

Reviewer: ```suggestion
When you upgrade the control plane on a cluster, the PKI certificates are automatically updated. Upgrading a cluster's Kubernetes version, whether a minor patch or a major release, results in renewed PKI certificates. This is the method that requires the least user actions when it comes to renewing PKI certificates. We recommend regularly updating your clusters to stay up-to-date with security fixes and best practices. By keeping your host cluster up-to-date, you prevent the scenario of PKI certificates from reaching their one-year expiration date.
```

To make this more concise.

Author: I'm declining this as we want to call out this method has the least actions.

<br />

<InfoBox>

> You can upgrade the Kubernetes version of a cluster by updating the Kubernetes layer of the [cluster profile](/cluster-profiles) and applying the cluster profile update to the cluster.

Reviewer 2: As line 48 talks about updating the pack version, here's another good link to add or to replace the cluster profile link: 

For guidelines on updating pack versions, review [Update the Pack Version](/cluster-profiles/task-update-profile#updatethepackversion).

I used it in the maintenance policy doc, which it talks about updating the pack version when your profile uses a disabled or deleted pack.

> You can upgrade the Kubernetes version of a cluster by updating the Kubernetes layer of the [cluster profile](/cluster-profiles) and applying the cluster profile update to the cluster.

Author: ```suggestion
You can upgrade the Kubernetes version of a cluster by updating the Kubernetes layer of the [cluster profile](/cluster-profiles) and applying the cluster profile update to the cluster. For guidelines on updating pack versions, review [Update the Pack Version](/cluster-profiles/task-update-profile#updatethepackversion).
```

</InfoBox>

# Manual Certificate Renewal

> You can renew the cluster PKI certificates on-demand using the Palette API or the Palette User Interface (UI). When you manually trigger a certificate renewal action, the action will result in cluster nodes becoming repaved. Palette will scale up the cluster control plane nodes count and deploy new nodes with new PKI certificates auto-generated by kubeadm. Once the new control plane nodes are active and available, Palette will remove the previous control plane nodes. The worker nodes renew once the control plane nodes are updated and available.

Reviewer 2: ```suggestion
You can renew the cluster PKI certificates on-demand using the Palette API or the Palette User Interface (UI). When you manually trigger a certificate renewal action, it results in cluster nodes becoming repaved. Palette will scale up the cluster control plane nodes count and deploy new nodes with new PKI certificates auto-generated by kubeadm. Once the new control plane nodes are active and available, Palette will remove the previous control plane nodes. The worker nodes renew once the control plane nodes are updated and available.
```

Reviewer 2: Is it "becoming repaved" or "being repaved"? Same question for line 61. I'm not sure if it makes any difference either - just wondered.

<br />

<WarningBox>

> A manual renewal of cluster PKI certificates will result in all cluster nodes becoming repaved.

Reviewer 2: ```suggestion
A manual renewal of cluster PKI certificates results in all cluster nodes becoming repaved.
```

</WarningBox>

> Review the following sections to learn more about manually renewing the cluster PKI certificates. Choose the preferred update method, API, or through the Palette UI.

Reviewer 2: ```suggestion
Review the following sections to learn more about manually renewing the cluster PKI certificates. Choose your preferred update method, using the Palette UI or the API.
```

Reviewer 2: Optionally, you could use a colon:

"...preferred update method: Palette UI or the API."

I switched the order to match the tab order.

<br />

<Tabs>

<Tabs.TabPane tab="UI" key="ui">

Use the following steps to update the cluster PKI certificates through the Palette UI.

## Prerequisites

> - A host cluster deployed.

Reviewer 2: ```suggestion
- A deployed host cluster.
```

- Access to the host cluster.

## Renew Cluster Certificate

1. Log in to [Palette](https://console.spectrocloud.com).

2. Navigate to the left **Main Menu** and select **Clusters**.

> 3. Select the host cluster you want to renew the PKI certificates.

Reviewer 2: ```suggestion
3. Select the host cluster to renew its PKI certificates.
```

Reviewer 2: If you want to keep "you want to", it's more correct to say the following, but then it sounds kind of formal:

"Select the host cluster for which you want to renew the PKI certificates."

4. From the cluster details page, click on **View K8s Certificates**.

  <br />

>   ![A view of the Palette UI with an arrow pointing to the k8s certificate button.](/clusters_cluster-management_certificate-management_cluster-details-page.png)

Reviewer 2: ```suggestion
  ![A view of the Palette UI with an arrow pointing to the **View K8s Certificates** button.](/clusters_cluster-management_certificate-management_cluster-details-page.png)
```

> 5. Next, select **Renew All** to start the renewal operation.

Reviewer 2: ```suggestion
5. Next, select **Renew All** to start the renewal process.
```

  <br />

  ![A view of the cluster certificates displaying the expiration date](/clusters_cluster-management_certificate-management_certificate-renew-page.png)

The renewal process may take several minutes, depending on the number of cluster nodes.

## Validate

> Using the following steps, you can validate that the cluster's PKI certificates got renewed.

Reviewer 2: ```suggestion
Using the following steps, you can validate that the cluster's PKI certificates were renewed.
```

<br />

1. Log in to [Palette](https://console.spectrocloud.com).

2. Navigate to the left **Main Menu** and select **Clusters**.

> 3. Select the host cluster you want to renew the PKI certificates.

Reviewer 2: ```suggestion
3. Select the host cluster with the renewed PKI certificates.
```

4. From the cluster details page, click on **View K8s Certificates**.

5. Review the expiration date for each component. Each component's expiration date will have a status of **365d** with a date that is one year away. 

</Tabs.TabPane>

<Tabs.TabPane tab="API" key="api">

Use the following steps to update the cluster PKI certificates through the Palette API.

## Prerequisites

- A Palette API key. You can learn how to create a Palette API key by reviewing the [API Key](/user-management/user-authentication/) documentation.

> - A host cluster deployed.

Reviewer 2: ```suggestion
- A deployed host cluster.
```

- Access to the host cluster.

## Renew Cluster Certificate

> 1. Set your Palette API key as an environment variable. Replace the placeholder `REPLACE_ME` with your actual API key.

Reviewer 2: ```suggestion
1. Set your Palette API key as an environment variable. Add your actual API key in place of `REPLACE_ME`.
```

  <br />

  ```shell
  export API_KEY=REPLACE_ME
  ```

> 2. Set the project ID as an environment variable. Replace the placeholder `REPLACE_ME` with your project ID. You can find the project ID on the Palette landing page. The project ID is displayed in the top right-hand corner of the page.

Reviewer 2: ```suggestion
2. Set the project ID as an environment variable. Add your project ID in place of `REPLACE_ME`. You can find the project ID on the Palette landing page. The project ID is displayed in the top right corner of the page.
```

  <br />

  ```shell
  export PROJECT_ID=REPLACE_ME
  ```

3. Set the cluster ID as an environment variable. Replace the placeholder `REPLACE_ME` with your cluster's ID. You can get the cluster ID from the cluster detail's page URL. The value after `clusters/` is the cluster ID.

  <br />

  ```shell
  export CLUSTER_ID=REPLACE_ME
  ```

4. Use the Palette API endpoint `https://api.spectrocloud.com/v1/spectroclusters/{uid}/k8certificates/renew` to renew a cluster's PKI certificates. The endpoint accepts the HTTP method `PATCH`, and the only required parameter is the ID of the cluster.

  <br />

  ```shell
  curl --request PATCH \
    --url  'https://api.spectrocloud.com/v1/spectroclusters/$CLUSTER_ID/k8certificates/renew' \
    --header 'Content-Type: application/json' \
    --header 'Accept: application/json' \
    --header "ApiKey: $API_KEY" \
    --header "ProjectUid: $PROJECT_ID"
  ```

3. No output is returned and an HTTP status `204` is expected. 

The renewal process may take several minutes, depending on the number of cluster nodes.

## Validate

Using the following steps, you can validate that the cluster's PKI certificates got renewed.

<br />

> 1. Set your Palette API key as an environment variable. Replace the placeholder `REPLACE_ME` with your actual API key.

Reviewer 2: ```suggestion
1. Set your Palette API key as an environment variable. Add your actual API key in place of `REPLACE_ME`. 
```

  <br />

  ```shell
  export API_KEY=REPLACE_ME
  ```

> 2. Set the project ID as an environment variable. Replace the placeholder `REPLACE_ME` with your project ID.

Reviewer 2: ```suggestion
2. Set the project ID as an environment variable. Add your project ID in place of `REPLACE_ME`.
```

  <br />

  ```shell
  export PROJECT_ID=REPLACE_ME
  ```

> 3. Set the cluster ID as an environment variable. Replace the placeholder `REPLACE_ME` with your cluster's ID. 

Reviewer 2: ```suggestion
3. Set the cluster ID as an environment variable. Add your cluster's ID in place of `REPLACE_ME`. 
```

  <br />

  ```shell
  export CLUSTER_ID=REPLACE_ME
  ```

4. Retrieve the cluster's certificate information from Palette by using the `https://api.spectrocloud.com/v1/spectroclusters/{uid}/k8certificates` endpoint.

  <br />

  ```
  curl \
   --url  'https://api.spectrocloud.com/v1/spectroclusters/$CLUSTER_ID/k8certificates' \
   --header 'Content-Type: application/json' \
   --header 'Accept: application/json' \
   --header "ApiKey: $API_KEY" \
   --header "ProjectUid: $PROJECT_ID"
  ```

> 5. Validate the output and confirm the expiration date is one year in the future.

Reviewer 2: ```suggestion
5. Validate the output and confirm the expiration date is one year away.
```

Reviewer 2: For consistency.

                      "name": "ca"
                  },
                  {
                      "certificates": [
                          {
                              "expiry": "2024-05-23T16:51:05.000Z",
                              "name": "kube-apiserver-etcd-client"
                          },
                          {
                              "expiry": "2024-05-23T16:51:05.000Z",
                              "name": "kube-etcd-healthcheck-client"
                          },
                          {
                              "expiry": "2024-05-23T16:51:05.000Z",
                              "name": "kube-etcd-peer"
                          },
                          {
                              "expiry": "2024-05-23T16:51:05.000Z",
                              "name": "kube-etcd-server"
                          }
                      ],
                      "expiry": "2033-05-23T16:45:22.209Z",
                      "name": "etcd-ca"
                  }
              ],
              "name": "ip-10-0-1-120.ec2.internal"
          }
      ]
  }
  ```

</Tabs.TabPane>

</Tabs>

<br />

# Advanced - Only Renew Control Plane Nodes

> You can configure Palette only to renew the PKI certificates for the control plane nodes. You can achieve this by using the annotation `spectrocloud.com/cert-renew-controlplane-only` and setting the value to `true`. To enable this behavior, you must use `kubectl` and apply the update to a Custom Resource Definition (CRD) created by Palette during the cluster deployment process.

Reviewer 2: ```suggestion
You can configure Palette to only renew the PKI certificates for the control plane nodes. You can achieve this by using the annotation `spectrocloud.com/cert-renew-controlplane-only` and setting the value to `true`. To enable this behavior, you must use `kubectl` and apply the update to a Custom Resource Definition (CRD) created by Palette during the cluster deployment process.
```

Use the following steps to configure Palette only to renew the certificates for control plane nodes.

## Prerequisites

- Kubectl is installed in your system.

- A host cluster deployed.

> - Access to the host cluster's kubeconfig file. Refer to [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide to learn how to use your cluster's kubeconfig file.

Reviewer 2: ```suggestion
- Access to the host cluster's kubeconfig file. Refer to the [Access Cluster with CLI](/clusters/cluster-management/palette-webctl) guide to learn how to use your cluster's kubeconfig file.
```

## Configure Cluster

> 1. Set your cluster name as an environment variable. Replace the placeholder `REPLACE_ME` with your cluster's name. 

Reviewer 2: ```suggestion
1. Set your cluster name as an environment variable. Add your cluster's name in place of `REPLACE_ME`. 
```

  <br />

  ```shell
  export CLUSTER_NAME=REPLACE_ME
  ```

1. Use the following command to retrieve the namespace of the CRD Palette created in your cluster.

  <br />

  ```shell
  namespace=$(kubectl get spc --all-namespaces --output jsonpath='{range .items[?(@.metadata.name=="'"$CLUSTER_NAME"'")]}{.metadata.namespace}{"\n"}{end}')
  ```

> 2. Use `kubectl` to update the CRD's definition to include the `spectrocloud.com/cert-renew-controlplane-only` annotation.

Reviewer 2: ```suggestion
2. Use `kubectl` to update the CRD to include the `spectrocloud.com/cert-renew-controlplane-only` annotation.
```

Reviewer 2: Removed because D in CRD is "definition".

  ```

3. Verify the annotation was set correctly with the command below. The expected output is `true`.

  <br />

  ```shell
  kubectl get spc/$CLUSTER_NAME --namespace $(kubectl get spc --all-namespaces --output jsonpath='{range .items[?(@.metadata.name=="'"$CLUSTER_NAME"'")]}{.metadata.namespace}{"\n"}{end}') --output jsonpath='{.metadata.annotations.spectrocloud\.com/cert-renew-controlplane-only}'
  ```

  Output
  ```
  true
  ```

4. Next, trigger a certificate renewal using either [Automatic Certificate Renewal](clusters/cluster-management/certificate-management#automaticcertificaterenewal) or [Manual Certificate Renewal](/clusters/cluster-management/certificate-management#manualcertificaterenewal).

The renewal process may take several minutes, depending on the number of cluster nodes.

## Validate

Using the following steps, you can validate that the cluster's PKI certificates are renewed only for the control plane nodes.

<br />

1. Log in to [Palette](https://console.spectrocloud.com).

2. Navigate to the left **Main Menu** and select **Clusters**.

> 3. Select the host cluster you want to renew the PKI certificates.

Reviewer 2: ```suggestion
3. Select the host cluster to renew its PKI certificates.
```
